pytorch_lightning/trainer/connectors/signal_connector.py

```diff
@@ -0,0 +1,108 @@
+import logging
+import os
+import signal
+import sys
+from signal import Signals
+from subprocess import call
+from types import FrameType, FunctionType
+from typing import Callable, List, Union
+
+import pytorch_lightning as pl
+from pytorch_lightning.utilities.imports import _fault_tolerant_training
+
+log = logging.getLogger(__name__)
+
+
+class HandlersCompose:
+    def __init__(self, signal_handlers: Union[List[Callable], Callable]):
+        if not isinstance(signal_handlers, list):
+            signal_handlers = [signal_handlers]
+        self.signal_handlers = signal_handlers
+
+    def __call__(self, signum: Signals, frame: FrameType) -> None:
+        for signal_handler in self.signal_handlers:
+            signal_handler(signum, frame)
+
+
+class SignalConnector:
+    def __init__(self, trainer: "pl.Trainer"):
+        self.trainer = trainer
+        self.trainer._terminate_gracefully = False
+
+    def register_signal_handlers(self) -> None:
+        sigusr1_handlers: List[Callable] = []
+        sigterm_handlers: List[Callable] = []
+
+        if _fault_tolerant_training():
+            sigusr1_handlers.append(self.fault_tolerant_sigusr1_handler_fn)
+
+        if self._is_on_slurm():
+            log.info("Set SLURM handle signals.")
+            sigusr1_handlers.append(self.slurm_sigusr1_handler_fn)
+
+        sigterm_handlers.append(self.sigterm_handler_fn)
+
+        if not self._is_on_windows():
+            if not self._has_already_handler(signal.SIGUSR1):
+                signal.signal(signal.SIGUSR1, HandlersCompose(sigusr1_handlers))
+
+            if not self._has_already_handler(signal.SIGTERM):
+                signal.signal(signal.SIGTERM, HandlersCompose(sigterm_handlers))
+
+    def slurm_sigusr1_handler_fn(self, signum: Signals, frame: FrameType) -> None:
+        if self.trainer.is_global_zero:
+            # save weights
+            log.info("handling SIGUSR1")
+            self.trainer.checkpoint_connector.hpc_save(self.trainer.weights_save_path, self.trainer.logger)
```

Comment:
Will this also handle weights properly that may be different on each rank? e.g. DeepSpeed, Sharded or Model Parallel?

Reply:
This is on Slurm. I guess we should explore this if we get access to it.